dkg-primitives/src/rounds/keygen.rs

```diff
@@ -0,0 +1,288 @@
+		trace!(target: "dkg", "🕸️  Getting outgoing keygen messages");
+
+		let keygen = &mut self.keygen;
+
+		if !keygen.message_queue().is_empty() {
+			trace!(target: "dkg", "🕸️  Outgoing messages, queue len: {}", keygen.message_queue().len());
+
+			let keygen_set_id = self.params.keygen_set_id;
+
+			let enc_messages = keygen
+				.message_queue()
+				.into_iter()
+				.map(|m| {
+					trace!(target: "dkg", "🕸️  MPC protocol message {:?}", m);
+					let serialized = serde_json::to_string(&m).unwrap();
+					return DKGKeygenMessage { keygen_set_id, keygen_msg: serialized.into_bytes() }
+				})
+				.collect::<Vec<DKGKeygenMessage>>();
+
+			keygen.message_queue().clear();
+			return enc_messages
+		}
+
+		vec![]
+	}
+
+	/// Handle incoming messages
+
+	fn handle_incoming(&mut self, data: DKGKeygenMessage, at: C) -> Result<(), DKGError> {
+		if data.keygen_set_id != self.params.keygen_set_id {
+			return Err(DKGError::GenericError { reason: "Keygen set ids do not match".to_string() })
+		}
+
+		let keygen = &mut self.keygen;
+
+		trace!(target: "dkg", "🕸️  Handle incoming keygen message");
+		if data.keygen_msg.is_empty() {
+			warn!(
+				target: "dkg", "🕸️  Got empty message");
+			return Ok(())
+		}
+
+		let msg: Msg<ProtocolMessage> = match serde_json::from_slice(&data.keygen_msg) {
+			Ok(msg) => msg,
+			Err(err) => {
+				error!(target: "dkg", "🕸️  Error deserializing msg: {:?}", err);
+				return Err(DKGError::GenericError {
+					reason: "Error deserializing keygen msg".to_string(),
+				})
+			},
+		};
+
+		if Some(keygen.party_ind()) != msg.receiver &&
+			(msg.receiver.is_some() || msg.sender == keygen.party_ind())
+		{
+			warn!(target: "dkg", "🕸️  Ignore messages sent by self");
+			return Ok(())
+		}
+		trace!(
+			target: "dkg", "🕸️  Party {} got message from={}, broadcast={}: {:?}",
+			keygen.party_ind(),
+			msg.sender,
+			msg.receiver.is_none(),
+			msg.body,
+		);
+		debug!(target: "dkg", "🕸️  State before incoming message processing: {:?}", keygen);
+		match keygen.handle_incoming(msg.clone()) {
+			Ok(()) => (),
+			Err(err) if err.is_critical() => {
+				error!(target: "dkg", "🕸️  Critical error encountered: {:?}", err);
+				return Err(DKGError::GenericError {
+					reason: "Keygen critical error encountered".to_string(),
+				})
+			},
+			Err(err) => {
+				error!(target: "dkg", "🕸️  Non-critical error encountered: {:?}", err);
+			},
+		}
+		debug!(target: "dkg", "🕸️  State after incoming message processing: {:?}", keygen);
+
+		Ok(())
+	}
+
+	/// Try finish
+
+	fn is_finished(&self) -> bool {
+		self.keygen.is_finished()
+	}
+
+	fn try_finish(mut self) -> Result<LocalKey<Secp256k1>, DKGError> {
+		info!(target: "dkg", "🕸️  Keygen is finished, extracting output, round_id: {:?}", self.params.round_id);
+		match self.keygen.pick_output() {
+			Some(Ok(key)) => {
+				info!(target: "dkg", "🕸️  local share key is extracted");
+				Ok(key)
+			},
+			Some(Err(e)) => panic!("Keygen finished with error result {}", e),
```

Comment:
is panicking here is desirable? I mean we return a result here, so we could just return another error, and it is up to the caller to handle these error instead of panicing here without any control.